### DISCUSSION TITLE
Android/mobile: Any way to detect spread/pinch gesture for zoom?

### DISCUSSION THREAD

**Discussion by uidbz** [fyne-io/fyne#4877] [Q&A]
Greetings,

Does anybody know of a way to detect spread/pinch gesture for zoom on mobile?

I am making an image viewer - on desktop I use the Scrolled(*fyne.ScrollEvent) function to implement zooming, but this is not suitable for mobile.

Any suggestions are most welcome :-)
Thank you.

**Comment by andydotxyz** [comment]
You can work with the touches directly by implementing the `fyne.Touchable` interface. We don't have the zoom abstraction coded into the library yet.

**Reply by Jacalz** [reply]
An API for handling touch gestures (both on mobile and desktops with touch screens) would be so cool. Touchpad gesture support on desktop would be a killer feature as well! :)

**Reply by andydotxyz** [reply]
The abstraction we have is designed capture the intent (as Scrolled mentioned above) - if you look at some of the early design documents the "Zoomed" was listed too, I figure that's what we should implement...

**Reply by Jacalz** [reply]
But gestures are not always scrolling. What I mean is a matter of gestures in the form of multiple fingers. I.e. touch pad gestures with three or more fingers in different directions. Also things like sliding in different direction to do a gesture on mobile. A more easy-to-use way to hook up pre defined gestures would be nice.

**Reply by andydotxyz** [reply]
I think we are taking across each other here. I wasn't suggesting we add more forms of scrolling. I was meaning we add new abstractions where the behaviour is clear.
For adding more feature types the mobile driver gives you access to the touch details. We have coded "gestures" in manually so others could too. Adding a gesture API specifically to make that easier may be interesting, but is mobile specific and for the most part I would rather we focus on standard behaviours across all platforms.

**Reply by uidbz** [reply]
Thank you very much for your comments :-)
I have looked at the fyne.Touchable interface, but it is not clear to me how multi-touch gestures can be achieved with it.

Do you by any chance have a small snippet illustrating this, or can point to a place in fyne where a multi-touch gesture has been implemented?

I have tried to look through the code, but it is a bit difficult for me to follow. I guess I need the 'tapId' that is used in various functions in fyne/internal/driver/mobile/canvas.go to differentiate the different touch events?

**Reply by andydotxyz** [reply]
Sorry there is not a demonstration of this, as you guessed the tapID is the index so you can track what different fingers are doing. It's similar to the iOS implementation so you might get some examples over there?
